src/hotspot/share/opto/loopnode.cpp

```diff
@@ -4765,6 +4765,19 @@ void PhaseIdealLoop::eliminate_useless_zero_trip_guard() {
       Node* opaque = head->is_canonical_loop_entry();
       if (opaque != nullptr) {
         useful_zero_trip_guard_opaques_nodes.push(opaque);
+#ifdef ASSERT
+        // See PhaseIdealLoop::do_unroll
+        // This property is desirable, but it maybe not hold after cloning a loop.
+        // In such a case, we bailout unrolling, and rely on IGVN to cleanup stuff.
+        // We are here before loop cloning (before iteration_split), so if there is a
+        // mess, it must come from the previous round of loop optimizations, which is bad.
+        // On the other hand, if this assert passes, bailing out in do_unroll means that
+        // this property was broken in the current round of loop optimization, which is
+        // acceptable.
+        if (!head->is_pre_loop() && !head->is_post_loop()) {
+          assert(opaque->outcnt() == 1 && opaque->in(1) == head->limit(), "IGVN should have cleaned that up!");
+        }
```

Comment:
I think we can just check `is_main_loop()` here. `is_canonical_loop_entry()` will bail out if we see anything else than main or post loops:

https://github.com/openjdk/jdk/blob/bdfd5e843a7d3db50edf4375e50449b0ce528f8a/src/hotspot/share/opto/loopnode.cpp#L6330-L6333


```suggestion
        if (head->is_main_loop()) {
          assert(opaque->outcnt() == 1 && opaque->in(1) == head->limit(), "IGVN should have cleaned that up!");
        }
```